File: test-fixtures/languages/java/sanitizer_test/Handlers.java
// Java sanitizer-fixture — parallel handlers per sink family.
package sanitizer_test;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import org.springframework.web.util.HtmlUtils;

public class Handlers {
    private Connection conn;

    // --- SQL injection ----------------------------------------------------

    public ResultSet sqlRaw(String userId) throws Exception {
        // SOURCE userId → string-built SQL → executeQuery (SINK).
        Statement stmt = conn.createStatement();
        return stmt.executeQuery("SELECT * FROM users WHERE id = '" + userId + "'");
    }

    public ResultSet sqlSafe(String userId) throws Exception {
        // SOURCE userId → PreparedStatement bind (SANITIZER) → executeQuery.
        PreparedStatement stmt = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        stmt.setString(1, userId);
        return stmt.executeQuery();
    }

    // --- XSS --------------------------------------------------------------

    public String xssRaw(String name) {
        return "<p>Hello, " + name + "</p>";
    }

    public String xssSafe(String name) {
        String safe = HtmlUtils.htmlEscape(name);
        return "<p>Hello, " + safe + "</p>";
    }
}

File: test-fixtures/languages/scala/sanitizer_test/Handlers.scala
// Scala sanitizer-fixture — parallel handlers per sink family.
package sanitizer_test

import java.sql.{Connection, PreparedStatement, ResultSet, Statement}
import org.springframework.web.util.HtmlUtils

class Handlers(conn: Connection) {

  def sqlRaw(userId: String): ResultSet = {
    val stmt: Statement = conn.createStatement()
    stmt.executeQuery(s"SELECT * FROM users WHERE id = '$userId'")
  }

  def sqlSafe(userId: String): ResultSet = {
    val stmt: PreparedStatement =
      conn.prepareStatement("SELECT * FROM users WHERE id = ?")
    stmt.setString(1, userId)
    stmt.executeQuery()
  }

  def xssRaw(name: String): String = s"<p>Hello, $name</p>"

  def xssSafe(name: String): String = {
    val safe = HtmlUtils.htmlEscape(name)
    s"<p>Hello, $safe</p>"
  }
}
